### ISSUE TITLE
Ltac2 anomaly with clear

### ISSUE THREAD

**Issue by herbelin** [coq/coq#14426]
#### Description of the problem

```
From Ltac2 Require Import Ltac2.  
Ltac2 g x := clear y; exact ($x).
Theorem foo (y:nat) : nat.
g constr:(y).
(* Error: Anomaly "in retyping: variable y unbound." *)
```

#### Coq Version

This is in 8.14+alpha. Coq 8.11 to 8.13 say instead "No such section variable or assumption: y".

**Comment by herbelin** [comment]
Another anomaly:

```
From Ltac2 Require Import Ltac2.
Ltac2 f () := let y := &y in intros y.
(* Anomaly "Missing hardwired primitive Ltac2.Control.hyp" *)
```
Also in 8.14+alpha but not in 8.11 to 8.13.

**Comment by ppedrot** [comment]
I can't reproduce this one. Are you sure you have the right files compiled and that it's not a build issue?

**Comment by herbelin** [comment]
Indeed, it seems to be related to an incomplete build. Strange that it accepted to `Require` w/o failing.

**Comment by SkySkimmer** [comment]
How is this not a regression?

Regardless the culprit is https://github.com/coq/coq/blob/525b63002402b53def5d1358b0c7c7f043e9b87d/user-contrib/Ltac2/tac2core.ml#L1476 (from c8a051cc9d85ef425b780a8454e8567e423171b4)
using protected_get_type_of from GlobEnv produces a non-anomaly `Error: Cannot reinterpret "y" in the current environment.`

**Comment by ppedrot** [comment]
I don't think we should turn the algebraic error into a user-readable one at that point. It has a cost, and instead we should register a printer in himsg. Incidentally, this also means that ill-typed Ltac1 terms will trigger the printer even when this is never displayed to the user.
